### ISSUE TITLE
no failover if vrrp_script times out

### ISSUE THREAD

**Issue by svenXY** [acassen/keepalived#535]
Hi,

my problem is that if the vrrp_script times out (here tested with sleep, but also with `select sleep(15)` in a mysql-failover situation), I want to handle that situation, but currently it does not even failover.

following setup:

2 mysqld servers with keepalived (tried v1.3.4 as well)

keepalived.conf (on both nodes)
```
! Configuration File for keepalived
global_defs {
      notification_email {
        admin@bla.laber
      }
      notification_email_from admin@bla.laber
      smtp_server localhost
      smtp_connect_timeout 30
      router_id mysql-ha
      }

vrrp_script check_mysql {
   script "sleep 10"
   interval 7
   timeout 3
   weight 2
   fall 3
   raise 3
}

vrrp_instance VI_1 {
      state BACKUP
      interface eth2
      virtual_router_id 51
      priority 100
      advert_int 1
      nopreempt  # only needed on higher priority node
      authentication {
      auth_type PASS
      auth_pass 1111
      }

      track_script {
        check_mysql
      }
      virtual_ipaddress {
        192.168.56.199/24 dev eth2 label eth2:1
      }
      #notify_master /mysql/keepalived_master.sh
      notify_backup "/usr/bin/logger -t Keepalived_xx Finished into state BACKUP"
      notify_master "/usr/bin/logger -t Keepalived_xx Finished into state MASTER"
}
```

To my understanding, when the script times out, keepalived should failover, but all I get is:

```
Feb 23 12:34:01 test-1 Keepalived_vrrp[9692]: Process [9701] didn't respond to SIGTERM
Feb 23 12:34:01 test-1 Keepalived_vrrp[9692]: pid 9701 exited due to signal 15
Feb 23 12:34:08 test-1 Keepalived_vrrp[9692]: Process [9704] didn't respond to SIGTERM
Feb 23 12:34:08 test-1 Keepalived_vrrp[9692]: pid 9704 exited due to signal 15
Feb 23 12:34:15 test-1 Keepalived_vrrp[9692]: Process [9708] didn't respond to SIGTERM
Feb 23 12:34:15 test-1 Keepalived_vrrp[9692]: pid 9708 exited due to signal 15
```

and the VIP is not switched to the other node.

Basically, keepalived switches properly when mysql is down and when keepalived gets stopped, but never when the check script times out.

I also tried to do it with trap and then exit 1 in the trap handler, but even that did not imporve on the situation.

I'm kind of lost here, but without a timeout handling, it's useless for us.

**Comment by pqarmitage** [comment]
The VRRP instance will transition to fault state, thereby causing the VIP to switch over, if the script returns a non-zero exit code, i.e. the script reports a failure. A timeout of the script is not considered a failure of the script, since it hasn't returned a non-zero exit code (it hasn't returned any exit code).

I think you will need to run a script that has it's own timeout, and returns an error if it times out. The following is an example:
```
#!/bin/bash

# With thanks to http://www.linuxjournal.com/content/sigalrm-timers-and-stdin-analysis

function timeout_handler
{
#	echo allowable time for execution exceeded.
	exit 1
}

function allow_time
{
    (
#	echo timer allowing $1 seconds for execution
	sleep $1
	kill -ALRM $$
    ) &
	child=$!
}

trap timeout_handler SIGALRM

allow_time ${1:-6}

# Now do what we really want, but it must be executed in background,
# and then waited on. Otherwise the trap handler won't run until
# the process terminates.
sleep ${2:-10} &

# Wait for process to terminate or alarm timer to expire
wait $!

# Clean up sub process
kill $child >/dev/null 2>&1

exit 0

```

**Comment by pqarmitage** [comment]
By the way, the message `Process [9701] didn't respond to SIGTERM` is due to a bug that looking into this caused me to find. keepalived is only allowing 2 microseconds for the child to terminate before it considers the child hasn't responded to the SIGTERM. I'll push a patch for that later.

Also, you should be getting a warning message in the log saying `WARNING - script 'sleep' resolved by path search '/bin/sleep'. Please specify fill path.` It would be better to specify the full path to the sleep executable in your config file.

And also, `raise` in your config should be `rise`.

**Comment by pqarmitage** [comment]
Commit 59bb31a has been pushed which resolves the `Process [9701] didn't respond to SIGTERM` messages.

I also found a bug in script name handling when it is not specified with an absolute path name, and that is resolved in commit ad317be.

**Comment by svenXY** [comment]
thanks for explanations and help! Much appreciated!

**Comment by ykfq** [comment]
Its a bug in keepalived-1.3.2.tar.gz. When I installed this version in January, 2017, I can't make the script "check_nginx.sh" work for chehking the status of nginx no matter what I did . And today, I suddeny found it has a new version keepalived-1.3.5.tar.gz and upgraded it, everything works fine now ! This is what it did:
`
wget http://www.keepalived.org/software/keepalived-1.3.5.tar.gz
tar -zxf keepalived-1.3.5.tar.gz 
service keepalived stop
./configure 
make && make install

service keepalived start
service nginx stop
netstat -ntlp | grep nginx
`
